### DISCUSSION TITLE
Synchronizing Zones in PowerDNS with MariaDB: Support for Primary-Primary Replication?

### DISCUSSION THREAD

**Discussion by MathZ31** [PowerDNS/pdns#14609] [Q&A]
Hello,

I am using PowerDNS with a MySQL (MariaDB) backend to manage my DNS zones. My goal is to ensure that any changes made to a zone on one PowerDNS server are automatically replicated to another PowerDNS server.

I have reviewed the PowerDNS documentation but couldn't find clear information regarding support for primary-primary (master-master) replication for DNS zones using a MySQL backend. Is it possible to configure PowerDNS to handle zone synchronization automatically via primary-primary database replication? If this is not directly supported, could you recommend the best approach to achieve this goal?

One option I am considering is setting up master-master replication for MariaDB. This would allow each MariaDB server to replicate changes made on the other server, ensuring that DNS records are synchronized between both PowerDNS servers. Do you think this approach would work well with PowerDNS? Are there any specific points or additional configurations to consider to ensure seamless integration between PowerDNS and MariaDB in a master-master replication setup?

Additionally, I would like to know if it's possible to have one server act as the primary and the other as a secondary while using master-master replication for MariaDB. How would this setup impact the operation of PowerDNS, and are there any best practices for implementing such a configuration?

For reference, here are the IP addresses of the servers involved in this setup:

PowerDNS Server 1: 10.144.0.43
PowerDNS Server 2: 10.144.0.44

Here are some details about my current configuration:

- PowerDNS configuration (pdns.conf):
```
api=yes
api-key=<api key>
include-dir=/etc/powerdns/pdns.d
launch=gmysql
gmysql-host=localhost
gmysql-port=3306
gmysql-dbname=powerdns
gmysql-user=powerdns
gmysql-password=<pwd>
gmysql-dnssec=yes
local-address=10.144.0.44
local-port=5300
loglevel=7
security-poll-suffix=
webserver=yes
webserver-address=10.144.0.44
webserver-allow-from=127.0.0.1,::1,10.144.0.0/16,10.190.0.0/16
webserver-hash-plaintext-credentials=yes
webserver-loglevel=detailed
webserver-max-bodysize=2
webserver-password=<pwd>
webserver-port=8081
webserver-print-arguments=yes
```

- PowerDNS recursor (recursor.conf)
```
allow-from=10.144.0.0/16, 127.0.0.0/8
api-key=<api key>
config-dir=/etc/powerdns
dnssec=validate
dont-query=127.0.0.0/8, 100.64.0.0/10, 169.254.0.0/16, 192.168.0.0/16, 172.16.0.0/12, ::1/128, fc00::/7, fe80::/10, 0.0.0.0/8, 192.0.0.0/24, 192.0.2.0/24, 198.51.100.0/24, 203.0.113.0/24, 240.0.0.0/4, ::/96, ::ffff:0:0/96, 100::/64, 2001:db8::/32
forward-zones=test.net=10.144.0.44:5300
hint-file=/usr/share/dns/root.hints
include-dir=/etc/powerdns/recursor.d
local-address=10.144.0.44
local-port=53
loglevel=7
lua-config-file=/etc/powerdns/recursor.lua
public-suffix-list-file=/usr/share/publicsuffix/public_suffix_list.dat
quiet=yes
security-poll-suffix=
webserver=yes
webserver-address=10.144.0.44
webserver-allow-from=127.0.0.1,::1,10.144.0.0/16,10.190.0.0/16
webserver-hash-plaintext-credentials=yes
webserver-loglevel=detailed
webserver-password=<pwd>
webserver-port=8082
```

- Database configuration (sample DNS records):
```
MariaDB [powerdns]> SELECT * FROM records;
+----+-----------+--------------------------+------+-----------------------------------------------------------+-------+------+----------+-----------+------+
| id | domain_id | name                     | type | content                                                   | ttl   | prio | disabled | ordername | auth |
+----+-----------+--------------------------+------+-----------------------------------------------------------+-------+------+----------+-----------+------+
|  2 |         2 | 0.144.10.in-addr.arpa    | SOA  | ns1.test.net root.test.net. 2024082309 28800 7200 604800 86400 | 86400 |    0 |        0 |           |    1 |
|  6 |         3 | test.net                 | SOA  | ns1.test.net root.test.net. 2024082309 28800 7200 604800 86400 | 86400 |    0 |        0 |           |    1 |
|  7 |         2 | 44.0.144.10.in-addr.arpa | PTR  | test.net                                                  | 86400 |    0 |        0 | 44        |    1 |
|  8 |         3 | test.net                 | A    | 10.144.0.44                                               | 86400 |    0 |        0 |           |    1 |
|  9 |         3 | test.net                 | NS   | ns1.test.net                                              | 86400 |    0 |        0 |           |    1 |
| 10 |         2 | 0.144.10.in-addr.arpa    | NS   | ns1.test.net                                              | 86400 |    0 |        0 |           |    1 |
+----+-----------+--------------------------+------+-----------------------------------------------------------+-------+------+----------+-----------+------+

```  
Thank you in advance for any suggestions or advice you can provide on this topic.

**Comment by Habbie** [comment]
> Are there any specific points or additional configurations to consider to ensure seamless integration between PowerDNS and MariaDB in a master-master replication setup?

In general (not specific to PowerDNS), it is wise to make sure all writes happen on the same instance, so writes cannot conflict with eachother. In the PowerDNS case, this is accomplished by only having primary=yes/secondary=yes enabled in one instance (or, in zero!), and by pointing all your API calls at one instance.

> Additionally, I would like to know if it's possible to have one server act as the primary and the other as a secondary while using master-master replication for MariaDB.

Do you mean primary/secondary in the DNS AXFR sense?
